x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SwapAliasesAndDeleteSourceIndexStep.java

```diff
@@ -0,0 +1,91 @@
+
+import java.util.Locale;
+
+/**
+ * Following a restore from snapshot operation this swaps all the aliases from the source index to the restored index and delete the
+ * source index.
+ */
+public class SwapAliasesAndDeleteSourceIndexStep extends AsyncActionStep {
+    public static final String NAME = "swap-aliases-to-restored";
+    private static final Logger logger = LogManager.getLogger(WaitForGreenIndexHealthStep.class);
+
+    private final String restoredIndexPrefix;
+
+    public SwapAliasesAndDeleteSourceIndexStep(StepKey key, StepKey nextStepKey, Client client, String restoredIndexPrefix) {
+        super(key, nextStepKey, client);
+        this.restoredIndexPrefix = restoredIndexPrefix;
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    @Override
+    public void performAction(IndexMetaData indexMetaData, ClusterState currentClusterState, ClusterStateObserver observer,
+                              Listener listener) {
+        String originalIndex = indexMetaData.getIndex().getName();
+        final String restoredIndexName = restoredIndexPrefix + originalIndex;
+        IndexMetaData restoredIndexMetaData = currentClusterState.metaData().index(restoredIndexName);
+
+        if (restoredIndexMetaData == null) {
+            String policyName = indexMetaData.getSettings().get(LifecycleSettings.LIFECYCLE_NAME);
+            String errorMessage = String.format(Locale.ROOT, "restored index [%s] doesn't exist. stopping execution of lifecycle [%s] for" +
+                " index [%s]", restoredIndexName, policyName, originalIndex);
+            logger.debug(errorMessage);
+            listener.onFailure(new IllegalStateException(errorMessage));
+            return;
+        }
+
+        // TODO extract and reuse - also see {@link org.elasticsearch.xpack.core.ilm.ShrinkSetAliasStep}
+        IndicesAliasesRequest aliasesRequest = new IndicesAliasesRequest()
+            .masterNodeTimeout(getMasterTimeout(currentClusterState))
+            .addAliasAction(IndicesAliasesRequest.AliasActions.removeIndex().index(originalIndex))
+            .addAliasAction(IndicesAliasesRequest.AliasActions.add().index(restoredIndexName).alias(originalIndex));
+        indexMetaData.getAliases().values().spliterator().forEachRemaining(aliasMetaDataObjectCursor -> {
+            AliasMetaData aliasMetaDataToAdd = aliasMetaDataObjectCursor.value;
+            // inherit all alias properties except `is_write_index`
+            aliasesRequest.addAliasAction(IndicesAliasesRequest.AliasActions.add()
+                .index(restoredIndexName).alias(aliasMetaDataToAdd.alias())
+                .indexRouting(aliasMetaDataToAdd.indexRouting())
+                .searchRouting(aliasMetaDataToAdd.searchRouting())
+                .filter(aliasMetaDataToAdd.filter() == null ? null : aliasMetaDataToAdd.filter().string())
+                .writeIndex(aliasMetaDataToAdd.writeIndex()));
```

Comment:
There's a comment above this about inheriting everything except for the writeIndex, but this looks like we're inheriting the write index setting?

Reply:
Ah, yes, I'm planning to refactor this into one component and reuse it in ShrinkSetAliasStep. It's not clear to me why we wouldn't want to inherit the write index (as we're deleting the original index in this one update). Do you know why we're skipping it in the shrink case?

Reply:
I don't think it's for a purposeful reason (at least I can't remember it being intentional)